lib/v8/weak.rb

```diff
@@ -39,7 +39,16 @@ def [](key)
         end
 
         def []=(key, value)
-          @values[key] = V8::Weak::Ref.new(value)
+          ref = V8::Weak::Ref.new(value)
+          ObjectSpace.define_finalizer(value, self.class.ensure_cleanup(@values, key, ref))
+
+          @values[key] = ref
+        end
+
+        def self.ensure_cleanup(values,key,ref)
+          proc {
```

Comment:
@novokshonovi  regarding the reasoning for the class method and rather awkward definition of finalizers, you need to be ultra careful with closures: see http://www.mikeperham.com/2010/02/24/the-trouble-with-ruby-finalizers/